packages/grid/x-data-grid/src/hooks/features/sorting/useGridSorting.ts

```diff
@@ -119,19 +119,18 @@ export const useGridSorting = (
    * API METHODS
    */
   const applySorting = React.useCallback<GridSortApi['applySorting']>(() => {
-    if (props.sortingMode === GridFeatureModeConstant.server) {
-      logger.debug('Skipping sorting rows as sortingMode = server');
-      apiRef.current.setState((state) => ({
-        ...state,
-        sorting: {
-          ...state.sorting,
-          sortedRows: gridRowIdsSelector(state, apiRef.current.instanceId),
-        },
-      }));
-      return;
-    }
-
     apiRef.current.setState((state) => {
+      if (props.sortingMode === GridFeatureModeConstant.server) {
```

Comment:
I moved everything into a single `setState` to make sure we always publish the `sortedRowsSet` event after